MobileCenterPush/MobileCenterPush/MSPush.m

```diff
@@ -55,10 +64,63 @@ - (instancetype)init {
     // Init channel configuration.
     _channelConfiguration = [[MSChannelConfiguration alloc] initDefaultConfigurationWithGroupId:[self groupId]];
     _appDelegate = [MSPushAppDelegate new];
+
+#if TARGET_OS_OSX
+    NSUserNotificationCenter *center = [NSUserNotificationCenter defaultUserNotificationCenter];
+
+    /*
+     * If there is a user notification center delegate already set by a customer before starting Push, assign the
+     * delegate to custom user notification center delegate.
+     */
+    if (center.delegate) {
+      userNotificationCenterDelegate = center.delegate;
+    }
+
+    // Set a delegate that will forward notifications to Push as well as a customer's delegate.
+    center.delegate = self;
+
+    // Observe delegate property changes.
+    [center addObserver:self
+             forKeyPath:kMSUserNotificationCenterDelegateKey
+                options:NSKeyValueObservingOptionNew
+                context:UserNotificationCenterDelegateContext];
+#endif
   }
   return self;
 }
 
+#if TARGET_OS_OSX
+- (void)dealloc {
+  [[NSUserNotificationCenter defaultUserNotificationCenter] removeObserver:self
+                                                                forKeyPath:kMSUserNotificationCenterDelegateKey
+                                                                   context:UserNotificationCenterDelegateContext];
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary *)change
+                       context:(void *)context {
+  if (context == UserNotificationCenterDelegateContext &&
+      [keyPath isEqualToString:kMSUserNotificationCenterDelegateKey]) {
+    userNotificationCenterDelegate = [change objectForKey:NSKeyValueChangeNewKey];
+    NSUserNotificationCenter *center = [NSUserNotificationCenter defaultUserNotificationCenter];
+    [center removeObserver:self forKeyPath:keyPath];
```

Comment:
What is the reason behind removing and re-adding self as observer?

Reply:
Ah got it, it's to avoid re-triggering the event for ourself. I would suggest testing `userNotificationCenterDelegate` for simplicity, if the new delegate is us just ignore?

Reply:
That's also possible but I don't think that can simplify the process. By removing/adding observer there can prevent going thru observer process that will eventually just ignore our delegate.

Reply:
I think it will at least improve readability.

Reply:
I agree with Clement as it improves readability.